server/src/main/java/org/elasticsearch/action/search/QueryPhaseResultConsumer.java

```diff
@@ -342,56 +361,71 @@ static long estimateRamBytesUsedForReduce(long size) {
         }
 
         public void consume(QuerySearchResult result, Runnable next) {
-            boolean executeNextImmediately = true;
-            synchronized (this) {
-                if (hasFailure() || result.isNull()) {
-                    result.consumeAll();
-                    if (result.isNull()) {
-                        SearchShardTarget target = result.getSearchShardTarget();
-                        emptyResults.add(new SearchShard(target.getClusterAlias(), target.getShardId()));
-                    }
-                } else {
-                    if (hasAggs) {
-                        long aggsSize = ramBytesUsedQueryResult(result);
-                        try {
-                            addEstimateAndMaybeBreak(aggsSize);
-                        } catch (Exception exc) {
-                            result.releaseAggs();
-                            releaseBuffer();
-                            onMergeFailure(exc);
-                            next.run();
-                            return;
+            if (hasFailure()) {
+                result.consumeAll();
+                next.run();
+            } else if (result.isNull()) {
+                result.consumeAll();
+                SearchShardTarget target = result.getSearchShardTarget();
+                SearchShard searchShard = new SearchShard(target.getClusterAlias(), target.getShardId());
+                synchronized (this) {
+                    emptyResults.add(searchShard);
+                }
+                next.run();
+            } else {
+                final long aggsSize = ramBytesUsedQueryResult(result);
+                boolean executeNextImmediately = true;
+                boolean hasFailure = false;
+                synchronized (this) {
+                    if (hasFailure()) {
+                        hasFailure = true;
```

Comment:
Isn't this check redundant? we are checking for failures at the start of the method

Reply:
Ah yea I wasn't ready to change the synchronisation behavior just yet on this one. It's very likely redundant but we used to check for failure under the lock before and I just didn't want to change that just yet. Let's remove more synchronisation in the follow-up IMO. This one we really should do in tiny steps or else :)

Reply:
For now I was happy to have the fast path for the common failure path if there's multiple results outstanding :)

Reply:
Make sense to me! Thanks for explaining 😄